tensorzero-internal/src/inference/providers/groq.rs

```diff
@@ -0,0 +1,2532 @@
+            Credential::None => Ok(GroqCredentials::None),
+            Credential::Missing => Ok(GroqCredentials::None),
+            _ => Err(Error::new(ErrorDetails::Config {
+                message: "Invalid api_key_location for Groq provider".to_string(),
+            })),
+        }
+    }
+}
+
+impl GroqCredentials {
+    pub fn get_api_key<'a>(
+        &'a self,
+        dynamic_api_keys: &'a InferenceCredentials,
+    ) -> Result<Option<&'a SecretString>, Error> {
+        match self {
+            GroqCredentials::Static(api_key) => Ok(Some(api_key)),
+            GroqCredentials::Dynamic(key_name) => {
+                Some(dynamic_api_keys.get(key_name).ok_or_else(|| {
+                    ErrorDetails::ApiKeyMissing {
+                        provider_name: PROVIDER_NAME.to_string(),
+                    }
+                    .into()
+                }))
+                .transpose()
+            }
+            GroqCredentials::None => Ok(None),
+        }
+    }
+}
+
+impl InferenceProvider for GroqProvider {
+    async fn infer<'a>(
+        &'a self,
+        request: ModelProviderRequest<'a>,
+        http_client: &'a reqwest::Client,
+        dynamic_api_keys: &'a InferenceCredentials,
+        model_provider: &'a ModelProvider,
+    ) -> Result<ProviderInferenceResponse, Error> {
+        let request_url = "https://api.groq.com/openai/v1/chat/completions".to_string();
+        let api_key = self.credentials.get_api_key(dynamic_api_keys)?;
+        let start_time = Instant::now();
+
+        let mut request_body =
+            serde_json::to_value(GroqRequest::new(&self.model_name, request.request)?).map_err(
+                |e| {
+                    Error::new(ErrorDetails::Serialization {
+                        message: format!(
+                            "Error serializing Groq request: {}",
+                            DisplayOrDebugGateway::new(e)
+                        ),
+                    })
+                },
+            )?;
+        let headers = inject_extra_request_data(
+            &request.request.extra_body,
+            &request.request.extra_headers,
+            model_provider,
+            request.model_name,
+            &mut request_body,
+        )?;
+
+        let mut request_builder = http_client
+            .post(request_url)
+            .header("Content-Type", "application/json");
+        if let Some(api_key) = api_key {
+            request_builder = request_builder.bearer_auth(api_key.expose_secret());
+        }
+
+        let raw_request = serde_json::to_string(&request_body).map_err(|e| {
+            Error::new(ErrorDetails::Serialization {
+                message: format!(
+                    "Error serializing request: {}",
+                    DisplayOrDebugGateway::new(e)
+                ),
+            })
+        })?;
+
+        let res = request_builder
+            .body(raw_request.clone())
+            .header(reqwest::header::CONTENT_TYPE, "application/json")
```

Comment:
this is still duplicated, see line 147. This caused an e2e test failure. Can you please remove the one on line 147?